File: app/src/main/java/codenevisha/ir/app/journey/Injection.java
package codenevisha.ir.app.journey;

import android.content.Context;
import android.support.annotation.NonNull;

import codenevisha.ir.app.journey.data.AppDataSource;
import codenevisha.ir.app.journey.data.AppRepository;
import codenevisha.ir.app.journey.data.local.HomeLocalDataSource;
import codenevisha.ir.app.journey.data.local.database.AppDatabase;
import codenevisha.ir.app.journey.data.remote.HomeRemoteDataSource;
import codenevisha.ir.app.journey.util.AppExecutors;


/**
 * Enables injection for
 * {@link AppDataSource} at compile time.
 * */

public class Injection {

  public static AppRepository provideApplicationRepository(@NonNull Context context) {

    AppDatabase database = AppDatabase.Companion.getInstance(context);

    return AppRepository.Companion.getInstance(
      HomeRemoteDataSource.Companion.getInstance(),
      HomeLocalDataSource.Companion.getInstance(new AppExecutors(), database.userDao())
    );

  }
}

File: app/src/main/java/codenevisha/ir/app/journey/app/Injection.kt
package codenevisha.ir.app.journey.app

import android.content.Context

import codenevisha.ir.app.journey.data.repository.HomeRepository
import codenevisha.ir.app.journey.data.impl.home.HomeRepositoryImpl
import codenevisha.ir.app.journey.data.impl.home.HomeLocalRepository
import codenevisha.ir.app.journey.data.local.database.AppDatabase
import codenevisha.ir.app.journey.data.impl.home.HomeRemoteRepository
import codenevisha.ir.app.journey.util.executer.AppExecutors


/**
 * Enables injection for
 * [HomeRepository] at compile time.
 */

object Injection {

    fun provideApplicationRepository(context: Context): HomeRepositoryImpl {

        val database = AppDatabase.getInstance(context)

        return HomeRepositoryImpl.getInstance(
                HomeRemoteRepository.instance,
                HomeLocalRepository.getInstance(AppExecutors(), database.userDao())
        )

    }
}
